File: app/src/main/java/nz/co/warehouseandroidtest/PermissionActivity.java
package nz.co.warehouseandroidtest;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import androidx.core.app.ActivityCompat;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

public class PermissionActivity extends AppCompatActivity {

    public static int PERMISSION_GRANTED = 0;
    public static int PERMISSION_DENIED = 1;

    public final String PACKAGE_URL_SCHEME = "package:";

    public static final String PERMISSION_EXTRA_FLAG = "nz.co.warehouseandroidtest.permission.extra_permission";

    public final int PERMISSION_REQUEST_CODE = 0;

    public boolean mIsRequiredPermissionCheck = false;

    private PermissionChecker checker = new PermissionChecker(this);

    public static void startActivityForResult(Activity activity, int requestCode, String[] permissions) {
        Intent intent = new Intent();
        intent.setClass(activity, PermissionActivity.class);
        intent.putExtra(PERMISSION_EXTRA_FLAG, permissions);
        ActivityCompat.startActivityForResult(activity, intent, requestCode, null);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getIntent() == null || !getIntent().hasExtra(PERMISSION_EXTRA_FLAG)) {
            throw new RuntimeException("PermissionsActivity needs to be started by static method startActivityForResult!");
        }

        setContentView(R.layout.activity_permission);

        mIsRequiredPermissionCheck = true;
    }

    @Override
    public void onResume() {
        super.onResume();

        if (mIsRequiredPermissionCheck) {
            if (checker.ifLackPermissions(getPermissions())) {
                requestPermissions(getPermissions());
            } else {
                allPermissionsGranted();
            }
        } else {
            mIsRequiredPermissionCheck = true;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (PERMISSION_REQUEST_CODE == requestCode && hasAllPermissionsGranted(grantResults)) {
            mIsRequiredPermissionCheck = true;
            allPermissionsGranted();
        } else {
            mIsRequiredPermissionCheck = false;
            showMissingPermissionDialog();
        }
    }

    private void showMissingPermissionDialog() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("Help");
        builder.setMessage("Current app lacks necessary permissions. \n\nPlease click \"Settings\" - \"Permission\" - to grant permissions. \n\nThen click back button twice to return.");

        builder.setNeutralButton("Quit", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                setResult(PERMISSION_DENIED);
                finish();
            }
        });

        builder.setPositiveButton("Settings", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                startAppSettings();
            }
        });

        builder.show();
    }

    public void startAppSettings() {
        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
        intent.setData(Uri.parse(PACKAGE_URL_SCHEME + getPackageName()));
        startActivity(intent);
    }

    private boolean hasAllPermissionsGranted(int[] grantResults) {
        for (int i = 0; i < grantResults.length; i++) {
            if (grantResults[i] == PackageManager.PERMISSION_DENIED) {
                return false;
            }
        }
        return true;
    }

    private void allPermissionsGranted() {
        setResult(PERMISSION_GRANTED);
        finish();
    }

    public void requestPermissions(String[] permissions) {
        ActivityCompat.requestPermissions(this, permissions, PERMISSION_REQUEST_CODE);
    }

    public String[] getPermissions() {
        return getIntent().getStringArrayExtra(PERMISSION_EXTRA_FLAG);
    }
}

File: app/src/main/java/nz/co/warehouseandroidtest/PermissionActivity.kt
package nz.co.warehouseandroidtest

import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat

class PermissionActivity : AppCompatActivity() {
    val PACKAGE_URL_SCHEME = "package:"
    val PERMISSION_REQUEST_CODE = 0
    var mIsRequiredPermissionCheck = false
    private val checker = PermissionChecker(this)
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (intent == null || !intent.hasExtra(PERMISSION_EXTRA_FLAG)) {
            throw RuntimeException("PermissionsActivity needs to be started by static method startActivityForResult!")
        }
        setContentView(R.layout.activity_permission)
        mIsRequiredPermissionCheck = true
    }

    public override fun onResume() {
        super.onResume()
        if (mIsRequiredPermissionCheck) {
            if (checker.ifLackPermissions(permissions)) {
                requestPermissions(permissions)
            } else {
                allPermissionsGranted()
            }
        } else {
            mIsRequiredPermissionCheck = true
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        if (PERMISSION_REQUEST_CODE == requestCode && hasAllPermissionsGranted(grantResults)) {
            mIsRequiredPermissionCheck = true
            allPermissionsGranted()
        } else {
            mIsRequiredPermissionCheck = false
            showMissingPermissionDialog()
        }
    }

    private fun showMissingPermissionDialog() {
        val builder = AlertDialog.Builder(this)
        builder.setTitle("Help")
        builder.setMessage("Current app lacks necessary permissions. \n\nPlease click \"Settings\" - \"Permission\" - to grant permissions. \n\nThen click back button twice to return.")
        builder.setNeutralButton("Quit") { dialog, which ->
            setResult(PERMISSION_DENIED)
            finish()
        }
        builder.setPositiveButton("Settings") { dialog, which -> startAppSettings() }
        builder.show()
    }

    fun startAppSettings() {
        val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
        intent.data = Uri.parse(PACKAGE_URL_SCHEME + packageName)
        startActivity(intent)
    }

    private fun hasAllPermissionsGranted(grantResults: IntArray): Boolean {
        for (i in grantResults.indices) {
            if (grantResults[i] == PackageManager.PERMISSION_DENIED) {
                return false
            }
        }
        return true
    }

    private fun allPermissionsGranted() {
        setResult(PERMISSION_GRANTED)
        finish()
    }

    fun requestPermissions(permissions: Array<String>) {
        ActivityCompat.requestPermissions(this, permissions!!, PERMISSION_REQUEST_CODE)
    }

    val permissions: Array<String>
        get() = intent.getStringArrayExtra(PERMISSION_EXTRA_FLAG)

    companion object {
        var PERMISSION_GRANTED = 0
        var PERMISSION_DENIED = 1
        const val PERMISSION_EXTRA_FLAG = "nz.co.warehouseandroidtest.permission.extra_permission"
        fun startActivityForResult(
            activity: Activity?,
            requestCode: Int,
            permissions: Array<String>
        ) {
            val intent = Intent()
            intent.setClass(activity, PermissionActivity::class.java)
            intent.putExtra(PERMISSION_EXTRA_FLAG, permissions)
            ActivityCompat.startActivityForResult(activity!!, intent, requestCode, null)
        }
    }
}